feat: described an ability to use async data in factory fields; refactored factories/base and some tests

nisemenov opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->
The main thing is the ability to use a coroutine type field for more complex logic (for example, using existing data from the DB).

## Description
- added some async methods (for both the main purpose and semantic)
- split some common logic into two separate methods **_get_initial_variables** and **_check_special_field**
- created **models.py** and **conftest.py** in SQLAlchemy test level
- replaced **create_tables** function with pytest fixture **fx_drop_create_meta**
- "pymongo<4.9" has been added in pyproject.toml to prevent
  ```
  PytestDeprecationWarning: Module 'beanie' was found, but when imported by pytest it raised: ImportError("cannot 
  import name '_QUERY_OPTIONS' from 'pymongo.cursor' (.venv/lib/python3.13/site-packages/pymongo/cursor.py)")
  ```
- appropriate tests and docs have been added

adhtruong commented:
Docs and updated examples look good

I think this is a good feature to have but would have concerns about
1. expanded interface and keeping these parts in sync for maintainability
2. potential edge case for this being a breaking change e.g. if `Coroutine` is already used it will now be resolved. This is probably a minor case

For 1, would it be possible to resolve coroutine from sync code itself by running in a thread or suggesting this in docs? This avoids pain of maintaining both sync and async versions at the expense of correctness and general proper async behaviour.

Something like 

```
import asyncio
import concurrent.futures

def run_coroutine_sync(coro):
    loop = asyncio.new_event_loop()
    with concurrent.futures.ThreadPoolExecutor() as executor:
        future = executor.submit(loop.run_until_complete, coro)
        return future.result()
```
with above being used when a coroutine is found as a field.

provinzkraut commented on polyfactory/factories/base.py:
```diff
@@ -396,6 +399,72 @@ def _get_or_create_factory(cls, model: type) -> type[BaseFactory[Any]]:
+        )
+
+    @classmethod
+    def _run_coroutine_sync(cls, field_value: Any) -> Any:
```
Not sure about this, as it could be really finnicky and likely to break in many edge cases. However, if we want this, it's probably better to use something more robust, like anyio's `BlockinPortal`: https://anyio.readthedocs.io/en/stable/threads.html#calling-asynchronous-code-from-an-external-thread

adhtruong replied:
@nisemenov @provinzkraut do you think it's more preferable to have this done on the user side instead?

My original suggestion here was unclear: technically could be done on user side this way without potentially async pitfalls and introducing a dependency for all users.

nisemenov replied:
I've just thought anyio's `BlockinPortal` is more flexible and safe. But I think leaving `ThreadPoolExecutor` is a good decision.

provinzkraut replied:
> @nisemenov @provinzkraut do you think it's more preferable to have this done on the user side instead?

User side would definitely preferable in general, yes

adhtruong replied:
Thanks both!

Some docs sound good. I'd favour trying to sign post to these packages rather than provide code snippets if possible to avoid maintenance of examples that may break in edge cases.

Some example of pre-fetching adapted from existing examples added may also be useful? E.g.  have this be called in top level function https://github.com/litestar-org/polyfactory/pull/641/files#diff-4be26d897af82d9f6c026da3c3386174e43b4a1b72940549a373e983becd3591R28 so async call is not needed for the factory method itself

nisemenov commented:
## Summary
- updated docs with describing an approach of using data async pre-fetched in factory fields;
- split some common logic into two separate methods **_get_initial_variables** and **_check_special_field**;
- created **models.py** and **conftest.py** in SQLAlchemy test level;
- replaced **create_tables** function with pytest fixture **fx_drop_create_meta**;
- "pymongo<4.9" has been added in pyproject.toml to prevent
  ```
  PytestDeprecationWarning: Module 'beanie' was found, but when imported by pytest it raised: ImportError("cannot 
  import name '_QUERY_OPTIONS' from 'pymongo.cursor' (.venv/lib/python3.13/site-packages/pymongo/cursor.py)")
  ```

adhtruong commented on polyfactory/factories/base.py:
```diff
@@ -396,6 +396,60 @@ def _get_or_create_factory(cls, model: type) -> type[BaseFactory[Any]]:
+        return result, generate_post, _build_context
+
+    @classmethod
+    def _check_special_field(
```
I'd prefer to revert this change. This isn't reusable and doesn't allow overriding properly and currently breaks if the factory field return None
